readthedocs/redirects/querysets.py

```diff
@@ -25,8 +25,55 @@ def api(self, user=None, detail=True):
             queryset = self._add_user_repos(queryset, user)
         return queryset
 
-    def get_redirect_path_with_status(self, path, language=None, version_slug=None):
-        for redirect in self.select_related('project'):
+    def get_redirect_path_with_status(self, path, full_path=None, language=None, version_slug=None):
+        # add extra fields with the ``path`` and ``full_path`` to perform a
+        # filter at db level instead with Python
+        queryset = self.annotate(
+            path=Value(
+                path,
+                output_field=CharField(),
+            ),
+            full_path=Value(
+                full_path,
+                output_field=CharField(),
+            ),
+        )
+        prefix = Q(
+            redirect_type='prefix',
+            path__startswith=F('from_url'),
+        )
+        page = Q(
+            redirect_type='page',
+            path__iexact=F('from_url'),
+        )
+        exact = (
+            Q(
+                redirect_type='exact',
+                from_url__endswith='$rest',  # Python implementation does "in"
```

Comment:
This doesn't seem like enough of a query. It isn't actually testing the URL at all is it, just confirming it ends with `$rest`? It seems to also need a `path__startswith=F('from_url')` or something?

Reply:
hrm... right. I'm not sure how to re-write this logic from Python into the query:

```
match = self.from_url.split('$rest')[0]
if full_path.startswith(match):
```

> It seems to also need a `path__startswith=F('from_url')` or something?

We would need to perform an operation over `from_url` field to split it first. I will try to find out this.

Reply:
This function can help us https://docs.djangoproject.com/en/2.2/ref/models/database-functions/#replace but it's on Django 2.1

Reply:
We should probably be able to combine a startswith and endswith, perhaps with an annotate that removes $rest or something. Seems complicated tho.

Reply:
We can probably do it with another queryset tho, with a filter on `$rest` and then a `startswith`.

Reply:
I pushed a possible solution using `Replace` from 2.2 --I'm not sure why it does not fully work yet.